### DISCUSSION TITLE
CRUD service response on unique constraint violation

### DISCUSSION THREAD

**Discussion by malta895** [mia-platform/community#175] [Q&A] [answered]
I need to match in a service the unique constraint violation error when inserting an item to a collection.

I tried to add a document that violates such constraint with a call like the following:
```bash
curl --location --request POST 'https://crud-service/some-collection/' \
--header 'Content-Type: application/json' \
--data-raw '{
    "someUniqueField": "some-already-existing"
}'
```

And I got the following error response:
```json
{
    "statusCode": 422,
    "error": "Unprocessable Entity",
    "message": "E11000 duplicate key error collection: some-collection index: some_unique_index dup key: { someUniqueField: \"some-already-existing\" }"
}
```

Is it guaranteed that a response with 422 status code is due to a unique constraint violation? 
Should I also parse the `message` in the response looking for the specific mongo error code `E11000`?

**Comment by frengooo** [comment, marked as answer]
Hi @malta895, looking at the source code of the `crud-service` seems like right now you can get a `422` only for this kind of error during an insert or bulk insert operation.

I don't know if, in the future, other cases could be added.

Since the eventual addition of a new `422` response status code is not a breaking change probably you should parse the `message`.

According to the [MongoDB Documentation](https://www.mongodb.com/docs/v4.2/core/index-unique/index.html), `E11000` is the "duplicate key" error, so it should be safe (for now at least).

**Reply by malta895** [reply]
Hi @silversoul93, thanks for your answer, I think  I'll parse the message to be sure.

However in my opinion the CRUD service should take charge of this parsing and return a more appropriate HTTP error status code such as `409 Conflict`. On the other hand this is a breaking change, so I'm not sure if it is the right thing to modify the current behavior.

**Reply by frengooo** [reply]
@fredmaggiowski @davidebianchi Maybe we can think to add some other `metadata` field that can be used for such use cases?
e.g.
```json
{
  "statusCode": 422,
  "error": "Unprocessable entity",
  "message": "E11000 ...",
  "metadata": {
    "crudServiceErrorId": "PRIMARY_KEY_INDEX_CONFLICT",
    "crudServiceErrorDescription": "..."
  }
}
```

This should not be breaking

**Reply by fredmaggiowski** [reply]
Hey, we could introduce the concept of `errorCodes` to provide fine graded details about the error (even though I like the 409 status code, we should investigate if it is appropriate).

Let's hear from the crud service maintainers @nicocampagna FYI

**Reply by TheRusCompass** [reply]
Hello guys! Thank you for all your suggestions!
Yes `errorCodes` is a very good idea! We were already thinking about it and we can add it to the roadmap.
At the moment I agree with @silversoul93 to parse the message
